File: server_elixir/test/afterlight/activities/chess_test.exs
defmodule Afterlight.Activities.ChessTest do
  use ExUnit.Case, async: true

  alias Afterlight.Activities.Chess

  defp move(from, to, promo \\ nil) do
    base = %{"type" => "move", "from" => from, "to" => to}
    if promo, do: Map.put(base, "promo", promo), else: base
  end

  defp at(board, sq) do
    <<file_c, rank_c>> = sq
    i = file_c - ?a + (rank_c - ?1) * 8
    String.at(board, i)
  end

  test "rulesVersion 1 is Afterlight house chess" do
    state = Chess.init_sim_state(slots: [0, 1])
    assert Chess.rules_version() == 1
    assert Chess.rules_name() == "afterlight-house-chess"
    assert state["rulesVersion"] == 1
    assert state["rulesName"] == "afterlight-house-chess"
    assert state["board"] == Chess.start_board()
    assert state["turn"] == "w"
    assert Chess.to_fen(state) == "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
  end

  test "out-of-turn and king-in-check moves are rejected; both boards stay identical" do
    a = Chess.init_sim_state()
    b = Chess.init_sim_state()
    assert Chess.same_position?(a, b)

    assert {:error, :out_of_turn} = Chess.apply_input(a, 1, move("e7", "e5"))
    assert Chess.same_position?(a, b)

    checked = Chess.from_fen("4k3/p7/8/4Q3/8/8/8/4K3 b - - 0 1")
    other = Chess.from_fen("4k3/p7/8/4Q3/8/8/8/4K3 b - - 0 1")
    assert checked["inCheck"] == true
    assert {:error, :king_in_check} = Chess.apply_input(checked, 1, move("a7", "a5"))
    assert Chess.same_position?(checked, other)
    assert checked["board"] == other["board"]
  end

  test "the same legal sequence keeps two boards on one position" do
    a = Chess.init_sim_state()
    b = Chess.init_sim_state()

    sequence = [{"e2", "e4"}, {"e7", "e5"}, {"g1", "f3"}, {"b8", "c6"}]

    {a, b} =
      Enum.reduce(sequence, {a, b}, fn {from, to}, {sa, sb} ->
        slot = if sa["turn"] == "w", do: 0, else: 1
        assert {:ok, na, _} = Chess.apply_input(sa, slot, move(from, to))
        assert {:ok, nb, _} = Chess.apply_input(sb, slot, move(from, to))
        assert Chess.same_position?(na, nb)
        {na, nb}
      end)

    assert at(a["board"], "e4") == "P"
    assert at(a["board"], "e5") == "p"
    assert Chess.same_position?(a, b)
  end

  test "castling kingside and queenside" do
    assert {:ok, cleared, _} =
             Chess.apply_moves(Chess.init_sim_state(), [
               move("e2", "e4"),
               move("e7", "e5"),
               move("g1", "f3"),
               move("b8", "c6"),
               move("f1", "e2"),
               move("g8", "f6")
             ])

    assert {:ok, castle, _} = Chess.apply_input(cleared, 0, move("e1", "g1"))
    assert castle["lastMove"]["castle"] == "kingside"
    assert at(castle["board"], "g1") == "K"
    assert at(castle["board"], "f1") == "R"
    assert at(castle["board"], "e1") == "."
    assert at(castle["board"], "h1") == "."
    refute String.contains?(castle["castling"], "K")
    refute String.contains?(castle["castling"], "Q")

    queen_side = Chess.from_fen("r3k2r/8/8/8/8/8/8/R3K2R w KQkq - 0 1")
    assert {:ok, oooo, _} = Chess.apply_input(queen_side, 0, move("e1", "c1"))
    assert oooo["lastMove"]["castle"] == "queenside"
    assert at(oooo["board"], "c1") == "K"
    assert at(oooo["board"], "d1") == "R"
  end

  test "en passant" do
    assert {:ok, ready, _} =
             Chess.apply_moves(Chess.init_sim_state(), [
               move("e2", "e4"),
               move("a7", "a6"),
               move("e4", "e5"),
               move("d7", "d5")
             ])

    assert ready["epSquare"] == "d6"
    assert {:ok, ep, _} = Chess.apply_input(ready, 0, move("e5", "d6"))
    assert ep["lastMove"]["enPassant"] == true
    assert at(ep["board"], "d6") == "P"
    assert at(ep["board"], "d5") == "."
    assert at(ep["board"], "e5") == "."
  end

  test "promotion defaults to queen and accepts underpromotion" do
    queen = Chess.from_fen("4k3/P7/8/8/8/8/8/4K3 w - - 0 1")
    assert {:ok, q, _} = Chess.apply_input(queen, 0, move("a7", "a8"))
    assert at(q["board"], "a8") == "Q"

    knight = Chess.from_fen("4k3/P7/8/8/8/8/8/4K3 w - - 0 1")
    assert {:ok, n, _} = Chess.apply_input(knight, 0, move("a7", "a8", "n"))
    assert at(n["board"], "a8") == "N"
  end

  test "checkmate (fool's mate)" do
    assert {:ok, result, _} =
             Chess.apply_moves(Chess.init_sim_state(), [
               move("f2", "f3"),
               move("e7", "e5"),
               move("g2", "g4"),
               move("d8", "h4")
             ])

    assert result["status"] == "complete"
    assert result["reason"] == "checkmate"
    assert result["winner"] == 1
    assert Chess.legal_moves(result) == []

    {stepped, ended} = Chess.step_simulation(result, [], 1)
    assert stepped["reason"] == "checkmate"
    assert {:match_ended, 1, %{"reason" => "checkmate"}} = ended
  end

  test "stalemate" do
    state = Chess.from_fen("7k/5Q2/6K1/8/8/8/8/8 b - - 0 1")
    assert state["status"] == "complete"
    assert state["reason"] == "stalemate"
    assert state["winner"] == nil
    assert state["inCheck"] == false
  end

  test "fifty-move draw after 100 quiet half-moves" do
    state = Chess.from_fen("4k3/8/8/8/8/8/8/4K3 w - - 99 1")
    assert state["status"] == "playing"
    assert {:ok, quiet, _} = Chess.apply_input(state, 0, move("e1", "e2"))
    assert quiet["halfmove"] == 100
    assert quiet["status"] == "complete"
    assert quiet["reason"] == "fifty_move"
    assert quiet["winner"] == nil
  end

  test "threefold repetition of the start position" do
    assert {:ok, result, _} =
             Chess.apply_moves(Chess.init_sim_state(), [
               move("g1", "f3"),
               move("g8", "f6"),
               move("f3", "g1"),
               move("f6", "g8"),
               move("g1", "f3"),
               move("g8", "f6"),
               move("f3", "g1"),
               move("f6", "g8")
             ])

    assert result["status"] == "complete"
    assert result["reason"] == "threefold"
    assert result["winner"] == nil
  end

  test "resignation awards the opponent" do
    state = Chess.init_sim_state()
    assert {:ok, white, _} = Chess.apply_input(state, 0, %{"type" => "resign"})
    assert white["status"] == "complete"
    assert white["reason"] == "resignation"
    assert white["winner"] == 1

    assert {:ok, black, _} = Chess.apply_input(Chess.init_sim_state(), 1, %{"type" => "resign"})
    assert black["winner"] == 0
  end

  test "agreed draw via offer and accept" do
    assert {:ok, opening, _} = Chess.apply_input(Chess.init_sim_state(), 0, move("e2", "e4"))
    assert {:ok, offered, _} = Chess.apply_input(opening, 1, %{"type" => "draw_offer"})
    assert offered["drawOffer"] == 1
    assert {:ok, accepted, _} = Chess.apply_input(offered, 0, %{"type" => "draw_accept"})
    assert accepted["reason"] == "agreement"
    assert accepted["winner"] == nil
  end
end

File: tests/chess.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import {
  CHESS_RULES_VERSION,
  CHESS_RULES_NAME,
  START_BOARD,
  initChessSimState,
  applyChessInput,
  applyChessMoves,
  legalMoves,
  samePosition,
  fromFen,
  toFen,
  stepChessSimulation,
} from '../shared/chessModel.js';

function play(state, slot, controls) {
  return applyChessInput(state, slot, controls);
}

function move(from, to, promo) {
  return { type: 'move', from, to, ...(promo ? { promo } : {}) };
}

test('chess rulesVersion 1 is Afterlight house chess', () => {
  const state = initChessSimState();
  assert.equal(CHESS_RULES_VERSION, 1);
  assert.equal(CHESS_RULES_NAME, 'afterlight-house-chess');
  assert.equal(state.rulesVersion, 1);
  assert.equal(state.rulesName, 'afterlight-house-chess');
  assert.equal(state.board, START_BOARD);
  assert.equal(state.turn, 'w');
  assert.match(toFen(state), /^rnbqkbnr\/pppppppp\/8\/8\/8\/8\/PPPPPPPP\/RNBQKBNR w KQkq - 0 1$/);
});

test('chess: out-of-turn and king-in-check moves are rejected; both boards stay identical', () => {
  const a = initChessSimState();
  const b = initChessSimState();
  assert.ok(samePosition(a, b));

  const blackFirst = play(a, 1, move('e7', 'e5'));
  assert.equal(blackFirst.ok, false);
  assert.equal(blackFirst.error, 'out_of_turn');
  assert.ok(samePosition(blackFirst.state, a));
  assert.ok(samePosition(blackFirst.state, b));

  const checked = fromFen('4k3/p7/8/4Q3/8/8/8/4K3 b - - 0 1');
  const other = fromFen('4k3/p7/8/4Q3/8/8/8/4K3 b - - 0 1');
  assert.equal(checked.inCheck, true);

  const ignored = play(checked, 1, move('a7', 'a5'));
  assert.equal(ignored.ok, false);
  assert.equal(ignored.error, 'king_in_check');
  assert.ok(samePosition(ignored.state, checked));
  assert.ok(samePosition(ignored.state, other));
  assert.equal(ignored.state.board, other.board);
});

test('chess: the same legal sequence keeps two boards on one position', () => {
  let a = initChessSimState();
  let b = initChessSimState();
  const sequence = [
    ['e2', 'e4'],
    ['e7', 'e5'],
    ['g1', 'f3'],
    ['b8', 'c6'],
  ];
  for (const [from, to] of sequence) {
    const slot = a.turn === 'w' ? 0 : 1;
    const ra = play(a, slot, move(from, to));
    const rb = play(b, slot, move(from, to));
    assert.equal(ra.ok, true);
    assert.equal(rb.ok, true);
    a = ra.state;
    b = rb.state;
    assert.ok(samePosition(a, b));
  }
  assert.equal(a.board[algebraicOffset('e4')], 'P');
  assert.equal(a.board[algebraicOffset('e5')], 'p');
});

function algebraicOffset(sq) {
  const file = 'abcdefgh'.indexOf(sq[0]);
  return (Number(sq[1]) - 1) * 8 + file;
}

test('chess: castling (kingside and queenside)', () => {
  const cleared = applyChessMoves(initChessSimState(), [
    { from: 'e2', to: 'e4' },
    { from: 'e7', to: 'e5' },
    { from: 'g1', to: 'f3' },
    { from: 'b8', to: 'c6' },
    { from: 'f1', to: 'e2' },
    { from: 'g8', to: 'f6' },
  ]);
  assert.equal(cleared.ok, true);
  const castle = play(cleared.state, 0, move('e1', 'g1'));
  assert.equal(castle.ok, true);
  assert.equal(castle.state.lastMove.castle, 'kingside');
  assert.equal(castle.state.board[algebraicOffset('g1')], 'K');
  assert.equal(castle.state.board[algebraicOffset('f1')], 'R');
  assert.equal(castle.state.board[algebraicOffset('e1')], '.');
  assert.equal(castle.state.board[algebraicOffset('h1')], '.');
  assert.ok(!castle.state.castling.includes('K'));
  assert.ok(!castle.state.castling.includes('Q'));

  const queenSide = fromFen('r3k2r/8/8/8/8/8/8/R3K2R w KQkq - 0 1');
  const oooo = play(queenSide, 0, move('e1', 'c1'));
  assert.equal(oooo.ok, true);
  assert.equal(oooo.state.lastMove.castle, 'queenside');
  assert.equal(oooo.state.board[algebraicOffset('c1')], 'K');
  assert.equal(oooo.state.board[algebraicOffset('d1')], 'R');
});

test('chess: en passant', () => {
  const ready = applyChessMoves(initChessSimState(), [
    { from: 'e2', to: 'e4' },
    { from: 'a7', to: 'a6' },
    { from: 'e4', to: 'e5' },
    { from: 'd7', to: 'd5' },
  ]);
  assert.equal(ready.ok, true);
  assert.equal(ready.state.epSquare, 'd6');
  const ep = play(ready.state, 0, move('e5', 'd6'));
  assert.equal(ep.ok, true);
  assert.equal(ep.state.lastMove.enPassant, true);
  assert.equal(ep.state.board[algebraicOffset('d6')], 'P');
  assert.equal(ep.state.board[algebraicOffset('d5')], '.');
  assert.equal(ep.state.board[algebraicOffset('e5')], '.');
});

test('chess: promotion defaults to queen and accepts underpromotion', () => {
  const queen = fromFen('4k3/P7/8/8/8/8/8/4K3 w - - 0 1');
  const q = play(queen, 0, move('a7', 'a8'));
  assert.equal(q.ok, true);
  assert.equal(q.state.board[algebraicOffset('a8')], 'Q');

  const knight = fromFen('4k3/P7/8/8/8/8/8/4K3 w - - 0 1');
  const n = play(knight, 0, move('a7', 'a8', 'n'));
  assert.equal(n.ok, true);
  assert.equal(n.state.board[algebraicOffset('a8')], 'N');
});

test('chess: checkmate (fool\'s mate)', () => {
  const result = applyChessMoves(initChessSimState(), [
    { from: 'f2', to: 'f3' },
    { from: 'e7', to: 'e5' },
    { from: 'g2', to: 'g4' },
    { from: 'd8', to: 'h4' },
  ]);
  assert.equal(result.ok, true);
  assert.equal(result.state.status, 'complete');
  assert.equal(result.state.reason, 'checkmate');
  assert.equal(result.state.winner, 1);
  assert.equal(legalMoves(result.state).length, 0);
  const stepped = stepChessSimulation(result.state, [], 1);
  assert.equal(stepped.matchEnded.reason, 'checkmate');
  assert.equal(stepped.matchEnded.winner, 1);
});

test('chess: stalemate', () => {
  const state = fromFen('7k/5Q2/6K1/8/8/8/8/8 b - - 0 1');
  assert.equal(state.status, 'complete');
  assert.equal(state.reason, 'stalemate');
  assert.equal(state.winner, null);
  assert.equal(state.inCheck, false);
});

test('chess: fifty-move draw after 100 quiet half-moves', () => {
  const state = fromFen('4k3/8/8/8/8/8/8/4K3 w - - 99 1');
  assert.equal(state.status, 'playing');
  const quiet = play(state, 0, move('e1', 'e2'));
  assert.equal(quiet.ok, true);
  assert.equal(quiet.state.halfmove, 100);
  assert.equal(quiet.state.status, 'complete');
  assert.equal(quiet.state.reason, 'fifty_move');
  assert.equal(quiet.state.winner, null);
});

test('chess: threefold repetition of the start position', () => {
  const shuffle = [
    { from: 'g1', to: 'f3' },
    { from: 'g8', to: 'f6' },
    { from: 'f3', to: 'g1' },
    { from: 'f6', to: 'g8' },
    { from: 'g1', to: 'f3' },
    { from: 'g8', to: 'f6' },
    { from: 'f3', to: 'g1' },
    { from: 'f6', to: 'g8' },
  ];
  const result = applyChessMoves(initChessSimState(), shuffle);
  assert.equal(result.ok, true);
  assert.equal(result.state.status, 'complete');
  assert.equal(result.state.reason, 'threefold');
  assert.equal(result.state.winner, null);
});

test('chess: resignation awards the opponent', () => {
  const state = initChessSimState();
  const whiteResigns = play(state, 0, { type: 'resign' });
  assert.equal(whiteResigns.ok, true);
  assert.equal(whiteResigns.state.status, 'complete');
  assert.equal(whiteResigns.state.reason, 'resignation');
  assert.equal(whiteResigns.state.winner, 1);

  const blackResigns = play(initChessSimState(), 1, { type: 'resign' });
  assert.equal(blackResigns.state.winner, 0);
});

test('chess: agreed draw via offer and accept', () => {
  const opening = play(initChessSimState(), 0, move('e2', 'e4'));
  const offered = play(opening.state, 1, { type: 'draw_offer' });
  assert.equal(offered.ok, true);
  assert.equal(offered.state.drawOffer, 1);
  const accepted = play(offered.state, 0, { type: 'draw_accept' });
  assert.equal(accepted.ok, true);
  assert.equal(accepted.state.reason, 'agreement');
  assert.equal(accepted.state.winner, null);
});
